Remove queries from the asset manager

Arnei opened the pull request:
Fixes #5966.

Removes all query related code from the asset manager. This means this asset manager cannot be accessed via queries anymore. Instead, the asset manager is exposing new methods which can be used instead.

What this boils down to is dumbing down the asset manager. The query language we used was powerful but complex. In removing it, we lose the power to create whatever query we want or need on a dime. Now, if there is something you want to do with the asset manager but there is no method for it, that method will have to be written first.
On the other hand, writing new methods should now be a lot easier. Named queries are less powerful, but comparatively easy to understand. Therefore accessing and working with the asset manager should be easier now.

This also allows us to get rid of the query-dsl dependency, which hopefully makes upgrading other depdencies easier.

### How to test this patch

Do whatever you can think of that would affect assets. Mostly this will be running workflows, but creativity is encouraged.


### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

Arnei committed:
Remove querydsl from assetmanager.impl.persistence

Replace it with namedQueries whereever possible.
Does not remove all query dsl due to dependencies in other packages
and because I can't get one of the queries quite right.

Arnei committed:
Use namedQueries instead of ASelectQueries in AssetManagerImpl

Some of the methods in the assetmanager do nothing but get snapshots
based on various arguments. Instead of using querydsl, this adds some
NamedQueries to SnapshotDto that can be used instead.

This is much more wordy than before, mostly due to many slightly different
JPQL queries (acceptable?) and code repetition for auth checking which
could definitely be more concise.

Arnei committed:
Remove AssetManagerImpl.getSnapshotsByDate(start, end)

There is a comment that claims that that function could lead to memory
overflow, and it is not used anywhere anyway, so this deletes it.

Arnei committed:
Remove query from index rebuild in asset manager

Arnei committed:
Get rid of QueryBuilder in Properties.java

Arnei committed:
Remove AQueryBuilder from AbstractAssetManagerRestEndpoint.java

Arnei committed:
Remove AQueryBuilder from AssetManagerStorageMoveOperationHandler.java

Arnei committed:
Remove AQueryBuilder from AssetManagerDeleteWorkflowOperationHandler.java

Arnei committed:
Remove AQueryBuilder from AssetManagerUpdatedEventHandler.java

Arnei committed:
Remove AQueryBuilder from AwsS3RestEndpoint.java

Arnei committed:
Remove AQueryBuilder from SelectVersionWorkflowOperationHandler.java

Arnei committed:
Remove AQueryBuilder from SchedulerServiceImpl.java

Arnei committed:
Remove AQueryBuilder from IndexServiceImpl.java

Arnei committed:
Remove AQueryBuilder from LiveScheduleServiceImpl.java

Arnei committed:
Remove AQueryBuilder from Workflows.java

Arnei committed:
Remove AQueryBuilder from OaiPmhUpdatedEventHandler.java

Arnei committed:
Remove AQueryBuilder from transcription services

Arnei committed:
Remove AQueryBuilder from DuplicateEventWorkflowOperationHandler.java

Arnei committed:
Remove AQueryBuilder from TestRestService.java

Arnei committed:
Remove dead code

Arnei committed:
Remove query building code from AssetManagerImpl and tests

Deletes more code related to query building from the asset manager.

Also deletes event listening functionality from the asset manager, because
it seemed pointless and was only used by unit tests anyway.

Deletes unit tests that tested the query architecture. Results in a lot of deleted
tests, as many tests were testing queries that were never used anywhere (but could
have been).
Adds some tests for our replacement methods, but not enough. In particular, there
are tests missing for some of the get methods, and the security testing needs to
be rethought altogether.

Arnei committed:
Rewrite saveProperty in asset manager db

Uses two db requests instead of one, so this is likely slower
than before. It is also a lot easier to understand what is going
on and works, so I'm content with it for now.

Arnei committed:
Remove the last remaining query dsl from AM db

No more "com.mysema.query.*" imports in Database.java.

Arnei committed:
Remove now unused query files from asset manager

Deletes query code from the asset manager that goes unused
due to the previous commits.

Arnei committed:
Remove the last of querydsl from the asset manager

Like the previous commit, removes a bunch of now unnecessary
code related to the query architecture. This finally allows
us to get rid of query dsl in the asset manager.

Arnei committed:
Remove querydsl from Opencast entirely

This should be the final step in making Opencast completely querydsl-free.
One less dependency to worry about.

Arnei committed:
Remove last remaining query classes from asset manager

With this all query related classes should be deleted from
the asset manager.

Arnei committed:
Checkstyle:ON in asset manager tests

Some files have checkstyle deactivated. There is no reason for this,
so let's activate it.

Arnei committed:
Fix-up asset manager select and security unit tests

Adds/Rewrites unit tests in the asset manager, specifically those
pertaining to getting from the asset manager and security checks.

Also tries to make the code somewhat more readable.

The most controversial change here is probably replacing how non-admin
users are authorized.

Arnei committed:
Merge branch 'develop' into murder-querydsl-from-the-assetmanager

# Conflicts:
#	modules/asset-manager-api/src/main/java/org/opencastproject/assetmanager/api/AssetManager.java

Arnei committed:
Fix issue with wrong uris

Apparently we need to potentially fix asset uris everytime
we get a snapshot from the database. Fun.

Arnei committed:
Also delete assets from asset store

The new delete methods for the asset manager were only
removing snapshots from the database, but left the associated
files on disk. This commit adds calls to the asset store in
the delete methods, which should result in the proper clean-up
of files.

Arnei committed:
Fix admin ui tests

Arnei committed:
Fix scheduler tests

Arnei committed:
Remove entwine imports from asset-manager-impl

Removing the query classes also got rid of most of the entwine
imports the asset manager was using. This replaces the last of
the entwine code that was still in use with more standard Java.

Arnei commented:
I'm failing to recreate the failing test in the workflows. I did create the test user with ROLE_STUDIO and ran the curl command, but it returns 200, which makes debugging difficult. @gregorydlogan do you maybe have any idea how I could better recreate this?

Edit: T'was probably to do with authentication. Rolled back my functional authentication changes.

Arnei committed:
Rollback usage of authorization.hasPermission

We cannot simply use the authorization service for our permission
checking apparently. Because this causing many hard-to-debug issues,
and rewriting how authorization is working in the asset manager is
*not* the goal of the PR this commit belongs to, this commit roles
back the usage of the authorization service in favor of using the
custom ACL check that was used before.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/SnapshotDto.java:
```diff
@@ -76,7 +76,161 @@
+            query = "SELECT s FROM Snapshot s "
+                + "WHERE s.mediaPackageId = :mediaPackageId "
+                + "AND (:organizationId IS NULL OR s.organizationId = :organizationId) "
+                + "AND s.version = ( "
+                + "  SELECT MAX(s2.version) FROM Snapshot s2 WHERE s2.mediaPackageId = s.mediaPackageId "
+                + ")"
```
Not really that important, but maybe this is simpler:
```sql
ORDER BY s.version DESC
LIMIT 1
```

Arnei replied:
Iirc I can't do that because JPQL does not support "LIMIT" like that.

I could instead add limit as a parameter to the related entity manager query, but I thought it would be bad style to split the logic of the query like that. I do not know which approach results in more efficient db queries.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/SnapshotDto.java:
```diff
@@ -76,7 +76,161 @@
+                + "AND s.archivalDate BETWEEN :startDate AND :endDate "
+                + "AND (:organizationId IS NULL OR s.organizationId = :organizationId)"
+        ),
+        @NamedQuery(
+            name = "Snapshot.findForIndexRebuild",
+            query = "SELECT s FROM Snapshot s "
+                + "WHERE s.version = ( "
+                + "  SELECT MAX(s2.version) FROM Snapshot s2 "
+                + "  WHERE s2.mediaPackageId = s.mediaPackageId "
+                + ") "
+                + "ORDER BY s.mediaPackageId DESC "
+        ),
```
<del>Not yet sure if that's a problem, but wouldn't this load all events at once from the database? Couldn't that lead to memory problems if there are 50000+ events?</del>

*Named query parameters in the JPA functions seem to limit this.*

Arnei replied:
And then contrary to my answer to the previous comment, here limit (and offset) are set in the entity manager query because they are supposed to be controlled ^^'.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/Database.java:
```diff
@@ -236,43 +242,64 @@ public void setAvailability(final VersionImpl version, final String mpId, final
-      var dtoOpt = Opt.nul(result).map(Full.fromTuple);
-      return dtoOpt.isSome() ? Optional.of(dtoOpt.get()) : Optional.empty();
+      return namedQuery.findOpt(
+              "Asset.findByChecksum",
```
This method is used only in a test which should probably use `findAssetByChecksumAndStoreAndOrg` instead. Getting rid of the test gets rid of this method and the new named query `Asset.findByChecksum`.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/AssetDto.java:
```diff
@@ -23,23 +23,68 @@
+        query = "SELECT a FROM Asset a "
+            + "INNER JOIN a.snapshot s "
+            + "WHERE a.checksum = :checksum"
+    ),
```
This is probably unnecessary since it's used only in one test. See comment at `findAssetByChecksum`.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/AssetDto.java:
```diff
@@ -23,23 +23,68 @@
+            + "INNER JOIN a.snapshot s "
+            + "WHERE a.checksum = :checksum"
+    ),
+    @NamedQuery(
+        name = "Asset.findByChecksumStorageIdAndOrganizationId",
+        query = "SELECT a FROM Asset a "
+            + "INNER JOIN a.snapshot s "
+            + "WHERE a.checksum = :checksum "
+            + "AND a.storageId = :storageId "
+            + "AND s.organizationId = :orgId "
+    ),
```
It seems weird to join only to filter and discard the data and then get the snapshot later anyway. See comment in `findAssetByChecksumAndStoreAndOrg`.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/Database.java:
```diff
@@ -348,16 +375,273 @@ public long countEvents(final String organization) {
-      var dtoOpt = Opt.nul(result).map(Full.fromTuple);
-      return dtoOpt.isSome() ? Optional.of(dtoOpt.get()) : Optional.empty();
+      return namedQuery.findOpt(
+              "Asset.findByChecksumStorageIdAndOrganizationId",
```
Looking at `Asset.findByChecksumStorageIdAndOrganizationId`, it joins the assets table with the snapshot table only to filter for one snapshot property (organization identifier) and then discard the snapshot data and return just an asset. What do we do with the asset? We use it to run another query to get the snapshot data :see_no_evil: 

If we need the join, can we use all the data we get?

Given that the checksum should usually be unique (check this assumption!), we could not join, but get the list of assets with the given checksum (should usually just be one), and use the first(?) one with a snapshot that belongs to the correct organization.

All of this is a bit weird, since `checksum` is not marked unique but is used as if it was (at least in combination with the organization).

Also, is it just me, or is `AssetDtos.Full` and `AssetDtos.Medium` a weird concept? Why not just use a tuple of `Asset` and `Snapshot` but copy over half of the snapshot data to another object?

lkiesow replied:
Actually :sweat_smile: … it seems like `AssetDtos.Full` is only ever used here in non-test scenarios. And the result of this method is only used to use snapshot data and the media package element identifier.. So, maybe we can return just that.

Arnei replied:
> Looking at Asset.findByChecksumStorageIdAndOrganizationId, it joins the assets table with the snapshot table only to filter for one snapshot property (organization identifier) and then discard the snapshot data and return just an asset. What do we do with the asset? We use it to run another query to get the snapshot data 🙈

I might be heavily misunderstanding what you're getting at, but the snapshot data is not discarded, is it? It is contained in the result and used in the `.map` right after this query is run. I don't see a second query to get the snapshot data.

> Also, is it just me, or is AssetDtos.Full and AssetDtos.Medium a weird concept? Why not just use a tuple of Asset and Snapshot but copy over half of the snapshot data to another object?

I honestly did not bother understanding what the purpose of all the ".Full" and ".Medium" classes was, so I can't speak on that :').

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/AssetDtos.java:
```diff
@@ -63,6 +40,14 @@ public static class Medium {
     private final String organizationId;
     private final String owner;
 
+    public Medium(AssetDto assetDto, Availability availability, String organizationId) {
```
The more I look at code related to this class the more I think we should just have named queries which then return a `Pair<Snapshot, Asset>` instead of building partial classes we then later convert, discard, …
Do we want to get rid of this entirely?

Arnei replied:
That is possibly a very good idea, but not something I wanted to address in this PR, for fear of losing sight of me goal of removing query-dsl and instead getting stuck trying to rewrite everything.

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/Database.java:
```diff
@@ -236,43 +242,64 @@ public void setAvailability(final VersionImpl version, final String mpId, final
+            Pair.of("mpId", mpId),
+            Pair.of("version", version.value()))
+          .apply(em)
+          .map(result -> new SnapshotDtos.Medium(
+              result,
+              Availability.valueOf(result.getAvailability()),
+              result.getStorageId(),
+              result.getOrganizationId(),
+              result.getOwner()
+          ));
```
Why? A `SnapshotDtos.Medium` is just a `Snapshot` with some data reordered? Why don't we just use `Snapshot`?

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/SnapshotDto.java:
```diff
@@ -76,7 +76,161 @@
-                + "where s.mediaPackageId = :mediaPackageId and s.organizationId = :organizationId")})
+                + "where s.mediaPackageId = :mediaPackageId and s.organizationId = :organizationId"),
+        @NamedQuery(
+            name = "Snapshot.findMediumByMpIdAndVersion",
```
```suggestion
            name = "Snapshot.findByMpIdAndVersion",
```

lkiesow commented on modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/persistence/Database.java:
```diff
@@ -348,16 +375,273 @@ public long countEvents(final String organization) {
+      }
+      if ("DESC".equals(orderByVersion)) {
+        namedQueryName = "Snapshot.findLatestVersionFirst";
+      }
```
We never promised that queries are unsorted if you don't pass any sort parameter → you can have one named query less:

```suggestion
      // Optional: check parameter
      if (!List.of("ASC", "DESC").contains(orderByVersion)) {
        // Exception here
      }

      String namedQueryName = "ASC".equals(orderByVersion) ? "Snapshot.findOldestVersionFirst" : "Snapshot.findLatestVersionFirst";
```

lkiesow commented on modules/scheduler-impl/src/test/java/org/opencastproject/scheduler/impl/SchedulerServiceImplTest.java:
```diff
@@ -852,78 +842,79 @@ public void testAddMultipleEventsConflict() throws Exception {
-      Assert.assertNotNull(e);
-    }
-  }
+// TODO: Figure out why getMediaPackage should throw an exception if it correctly returns an existing mediapackage
```
If we can't figure that out: Drop the test. It wouldn't be our first test with errors in the test.
In any case, we shouldn't just comment this out.

lkiesow reviewed (changes requested):
That you for doing this. I hated the review! But that you nonetheless :D 

This generally looks good. Most things I found probably existed before. I also did some local testing where I ran this by processing a few videos → no problems.

I'm going to do a few additional tests with different users in the next days.

Arnei committed:
Remove Asset.findByChecksum from asset manager

This method is currently going unused and has been deemed
unnecessary, so this commit deletes it.

Arnei committed:
Simplify name of named query

Arnei committed:
No unordered result for getSnapshots

As @lkiesow rightly mentioned, we don't need to promise that
queries are unsorted if no sort parameter is passed. Therefore
we can simplify this and get rid of a named query.

Arnei committed:
Remove assetManager.getMediaPackage test from scheduler

I could not figure out why this test makes any sense.
Instead of leaving it there as a massive comment block,
this removes it.

Arnei committed:
Merge branch 'develop' into murder-querydsl-from-the-assetmanager

Arnei committed:
Merge branch 'develop' into murder-querydsl-from-the-assetmanager

Arnei committed:
Merge branch 'develop' into murder-querydsl-from-the-assetmanager

Arnei committed:
Remove unused dependency

lkiesow reviewed (approved):
Tested this again and it looks good.

Ignoring `AssetDtos.Full` and `AssetDtos.Medium` is fair since it's not an issue created by this pull request. I'll create an OSDN issue for that :grin:
